src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -401,16 +416,21 @@ public async Task<uint256> Rewind()
                 if (this.blockHash == this.innerBlockHash)
                     this.unspents.Clear();
 
-                if (this.unspents.Count != 0)
+                // Check if rewind data is available in local cache. If it is 
+                // we can rewind and there is no need to check underlying storage.
+                if (this.cachedRewindDataList.Any())
                 {
-                    // More intelligent version can restore without throwing away the cache. (as the rewind data is in the cache).
-                    this.unspents.Clear();
-                    this.blockHash = this.innerBlockHash;
+                    RewindData lastRewindData = this.cachedRewindDataList.Last();
+
+                    this.RemoveTransactions(lastRewindData);
+                    this.RestoreOutputs(lastRewindData);
 
-                    this.logger.LogTrace("(-)[REWOUND_TO_INNER]:'{0}'", this.blockHash);
-                    return this.blockHash;
+                    this.cachedRewindDataList.Remove(lastRewindData);
+                    this.logger.LogTrace("(-)[REMOVED_FROM_BATCH]:'{0}'", lastRewindData.PreviousBlockHash);
+                    return lastRewindData.PreviousBlockHash;
```

Comment:
For example here, I don't understand, why don't we need to assign new value into this.blockHash